src/Microsoft.Azure.WebJobs.Host/Blobs/Listeners/BlobListenerFactory.cs

```diff
@@ -198,50 +197,103 @@ internal class BlobListenerFactory : IListenerFactory
+                return singletonBlobListener;
             }
             else
             {
                 // We've already created the singleton blob listener
                 // so just return our queue listener. Note that while we want the
                 // blob scan to be singleton, the shared queue listener needs to run
                 // on ALL instances so load can be scaled out
-                return queueListener;
+                return new NullListener();
             }
         }
 
         private async Task RegisterWithSharedBlobListenerAsync(
             string hostId,
             SharedBlobListener sharedBlobListener,
             IStorageBlobClient blobClient,
-            IStorageQueue hostBlobTriggerQueue,
-            IMessageEnqueuedWatcher messageEnqueuedWatcher,
+            DispatchProcessor dispatchProcessor,
             CancellationToken cancellationToken)
         {
             BlobTriggerExecutor triggerExecutor = new BlobTriggerExecutor(hostId, _functionId, _input,
-                BlobETagReader.Instance, new BlobReceiptManager(blobClient),
-                new BlobTriggerQueueWriter(hostBlobTriggerQueue, messageEnqueuedWatcher));
+                BlobETagReader.Instance, new BlobReceiptManager(blobClient), dispatchProcessor);
 
             await sharedBlobListener.RegisterAsync(_container, triggerExecutor, cancellationToken);
         }
 
-        private void RegisterWithSharedBlobQueueListenerAsync(
-            SharedBlobQueueListener sharedBlobQueueListener,
-            IStorageBlobClient blobClient,
-            IStorageQueueClient queueClient)
+        private class BlobMessageHandler : MessageHandler
         {
-            BlobQueueRegistration registration = new BlobQueueRegistration
+            private readonly IStorageBlobClient _blobClient;
+            private readonly ITriggeredFunctionExecutor _executor;
+            private readonly IBlobWrittenWatcher _blobWrittenWatcher;
+            internal BlobMessageHandler(ITriggeredFunctionExecutor executor,
+                                        IStorageBlobClient blobClient,
+                                        IStorageQueueClient queueClient,
+                                        IBlobWrittenWatcher blobWrittenWatcher)
             {
-                Executor = _executor,
-                BlobClient = blobClient,
-                QueueClient = queueClient
-            };
+                _executor = executor;
+                _blobClient = blobClient;
+                _blobWrittenWatcher = blobWrittenWatcher;
+                PoisonQueue = queueClient?.GetQueueReference(HostQueueNames.BlobTriggerPoisonQueue).SdkObject;
+            }
+            public override CloudQueue PoisonQueue { get; }
 
-            sharedBlobQueueListener.Register(_functionId, registration);
+            public override async Task<FunctionResult> TryExecuteAsync(JObject data, CancellationToken cancellationToken)
```

Comment:
This is basically the same logic that previously existed in BlobQueueTriggerExecutor, right? Where is the logic that the functionID is still valid (function still exists)?

Reply:
are you referring to this code
https://github.com/Azure/azure-webjobs-sdk/blob/e648136df61a6f43ac7bc139739e24df9f2fbfef/src/Microsoft.Azure.WebJobs.Host/Blobs/Listeners/BlobQueueTriggerExecutor.cs#L52-L70

it is moved to sharedQ, it will be applied for every message before dispatch to their handlers